{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lesson 32: Basic Local Alignment Search Tool (BLAST)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*This tutorial was generated from a Jupyter notebook.  You can download the notebook [here](l32_blast.ipynb).*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/saladi/anaconda3/lib/python3.4/site-packages/Bio/SearchIO/__init__.py:211: BiopythonExperimentalWarning: Bio.SearchIO is an experimental submodule which may undergo significant changes prior to its future official release.\n",
      "  BiopythonExperimentalWarning)\n"
     ]
    }
   ],
   "source": [
    "import Bio.Entrez\n",
    "import Bio.SeqIO\n",
    "import Bio.Blast.NCBIWWW\n",
    "import Bio.SearchIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Background and Theory\n",
    "\n",
    "In the process of understanding our place in the world, we often seek to ask whether DNA or proteins that we know or work with bear some resemblance to other molecules in the universe or, a perhaps a specific subset, of the sequence space. This problem is potentially enormous given the size of the sequence space (billions of letters) versus the size of the query (hundreds of letters).\n",
    "\n",
    "The most widely used algorithm/solution to this problem is BLAST.\n",
    "\n",
    "Given the size of the sequence space and the potential matches, we need an intelligent way to search. BLAST provides an efficient algorithm to do this explained below. See this [blog post](http://blog.thegrandlocus.com/2014/06/once-upon-a-blast) for a nice retrospective on the development of BLAST."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Algorithm\n",
    "\n",
    "1. Filter the query sequence for regions of low complexity. NCBI describes these regions [here](http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=FAQ#LCR) as :\n",
    "\n",
    "    >Regions with low-complexity sequence have an unusual composition that can create problems in sequence similarity searching. Low-complexity sequence can often be recognized by visual inspection. For example, the protein sequence PPCDPPPPPKDKKKKDDGPP has low complexity and so does the nucleotide sequence AAATAAAAAAAATAAAAAAT. Filters are used to remove low-complexity sequence because it can cause artifactual hits.\n",
    "    \n",
    "    >In BLAST searches performed without a filter, high scoring hits may be reported only because of the presence of a low-complexity region. Most often, it is inappropriate to consider this type of match as the result of shared homology. Rather, it is as if the low-complexity region is \"sticky\" and is pulling out many sequences that are not truly related.\n",
    "\n",
    "    >For amino acid queries this compositional bias is determined by the SEG program (Wootton and Federhen, 1996). For nucleotide queries it is determined by the DustMasker program (Morgulis, et al., 2006).\n",
    "\n",
    "2. The query sequence is transformed into subsequences of length $k$ \"k-words\". The default $k$ is 3 for protein queries and 11 for nucleotide queries.\n",
    "\n",
    "3. Each k-word is compared to all of the k-words found in the database of sequences via a substitution matrix. The high-scoring database words for each k-word are organized into an efficient search tree. This is performed for each of the k-words in the query sequence.\n",
    "\n",
    "    >How do we do comparisions?  This is an involved discussion of substitution matrices.  We may do this in class, but won't delve into it here.\n",
    "\n",
    "4. Sequences with exact matches of high scoring words are then retrieved from the database. Since each high-scoring word is unique and hashable, you can imagine a data structure/organization like a Python dictionary can be used to retrieve these matches with low computational cost.\n",
    "6. For sequences with hits of the each high scoring word, the alignment with the query sequence is extended in either direction until the overall score starts to decrease. These are called high scoring pairs (HSPs).\n",
    "7. HSPs with a score above a given cutoff are retained. This cutoff is empirically determined by computing alignment scores of many random sequences. \n",
    "8. The significance of each high scoring pair is computed via the [Gumbel extreme value distribution](https://en.wikipedia.org/wiki/Gumbel_distribution). Ungapped Smith-Waterman alignment scores between two random sequences follow the Gumbel EVD; this has not been proven for the case of gapped alignments (whether step 6 uses a gapped or ungapped alignment depends on parameters specified to BLAST).\n",
    "\n",
    "    >$$p \\left( S \\ge x \\right)=1-\\exp\\left( -e^{ -\\lambda \\left( x-\\mu \\right) } \\right)$$\n",
    "    \n",
    "    ><center> where </center>\n",
    "    \n",
    "    >$$\\mu =\\frac{ \\log \\left( Km'n' \\right) }{\\lambda }\\;$$\n",
    "\n",
    "    >The probability $p$ that a score $S$ is greater than $x$ is a function of parameters $K$, $\\lambda$, $m'$, and $n'$. $\\lambda$ and $K$ depend upon the substitution matrix, gap penalties, and sequence composition (the letter frequencies) of the database and are found by fitting to the Grumbel EVD. $m'$ and $n'$ are the effective lengths of the query and HSP sequences, i.e. adjusting for the \"edge effect\" that an alignment at the end of a sequence is less likely to have enough letters to build an optimal alignment. $H$ is the average expected score per aligned pair of residues in an alignment of two random sequences.\n",
    "    \n",
    "    >$$m'\\approx m-\\frac{\\ln Kmn }{H}$$\n",
    "    \n",
    "    >$$n'\\approx n-\\frac{\\ln Kmn}{H}$$\n",
    "    \n",
    "    > What we'd like to know, however, is how likely the alignment generated was found by chance. To do so, we want to find the expectation, or the average outcome, of the distribution. The expect score $E$ does just that; it gives the number of times that an unrelated sequence in the database of size $D$ would obtain a score $S$ higher than $x$.\n",
    "\n",
    "    > $$E\\approx 1-e^{-p\\left( s>x \\right)D}$$\n",
    "    \n",
    "    > When $p<0.1$, $E$ is approximated by the Poisson distribution as\n",
    "    \n",
    "    > $$E\\approx pD$$\n",
    "\n",
    "    >The expect score $E$ is often referred to as the \"E-value\" and the score $s$ as the \"bit score\". Critically:\n",
    "    \n",
    "    > - the score $s$ is independent of the size of the database and has everything to do with the quality of the alignment\n",
    "    > - the expect score is completely dependent on the size of the database\n",
    "9. HSPs are combined into longer alignments where possible and the scores are recomputed.\n",
    "10. Finally, those alignments with an E-value less than user-specified threshold are reported.\n",
    "\n",
    "Remember, it is the **bit score** that is independent of the  database size, so, an important point:\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "<center>When comparing hits from queries performed on different databases, always use the **bit-score** not the **E-value**!</center>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### BLAST programs:\n",
    "\n",
    "|Query|Database|Name|Description|\n",
    "|----|----------|-----------|\n",
    "|Nucleotide|Nucleotide|blastn|Nucleotide against nucleotide database|\n",
    "|Protein|Protein|blastp|Protein against protein database|\n",
    "|Protein|Protein|Position-Specific Iterative BLAST (PSI-BLAST)|Useful to find distant relatives of a *protein*. A list of closely related proteins is created via an initial BLASTp search and then aligned into a \"profile\" sequence, which summarises the significant features and filters out less significant ones. This profile is then run against the database again producing a larger set of proteins. This larger set is used to construct another profile. The process is \"iterated\". PSI-BLAST is much more sensitive in detecting up distant relationships than a standard BLASTp.|\n",
    "|Nucleotide (6-frame translation)|Protein|blastx|All six-frame frames of the nucleotide query sequence, i.e. +1,+2,+3,-1,-2,-3, are queried against a protein sequence database.|\n",
    "|Nucleotide (6-frame translation)|Nucleotide (6-frame translation)|tblastx| All translations of the query sequences are compared to all translations of the sequence database. tblastx is to useful in detecting very distant relationships between nucleotide sequences. Slowest of the BLAST family.|\n",
    "|Protein|Nucleotide (6-frame translation)|tblastn| Compares a protein query against all translations of a nucleotide sequence database. Good for searching a draft genome for proteins of interest.|\n",
    "|Any (Many query sequences)|Any (One database)|megablast|When querying large numbers of input sequences via the same sequence database, there are a number of steps that can be peformed together and decrease overall computational time. \"megablast\" does just this and is much faster than running standard BLAST multiple times. Simply put, it concatenates the query sequences to form a single long query, searches this new query against the BLAST database, and then analyzes the search results to generate individual alignments and scores.|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BLAST flavors:\n",
    "\n",
    "Many software pacakges have been written for BLAST exploiting fine and/or coarse-grained parallelism in the algorithm. For example\n",
    "- Instead of comparing one k word to those in a database sequentiallially, these operations can be processed in parallel\n",
    "- Instead of comparing high scoring words sequentially, this can be done in parallel\n",
    "- Instead of searching a single large database, it can be broken up and done in parallel\n",
    "- Instead of aligning many similar sequences through different HSPs, similar sequences can be collapsed into a smaller database that is instead searched with the query sequence\n",
    "\n",
    "Primarily, depending on the level/type of parallellism exploited together with special hardward architecture, several software packages have been written. Some take advantage of multiple processing cores on a single machine (PTHREADS) or multiple nodes in a cluster (MPI) and others leverage special hardware like graphical processing units (GPUs) and  field-programmable gate arrays (FPGAs). See a more complete list [here](https://en.wikipedia.org/wiki/BLAST#Accelerated_versions).\n",
    "\n",
    "Compressively-accelerated BLAST [(CaBLAST)](http://cast.csail.mit.edu/) has a pretty cool approach to increasing the speed of BLAST searches. Sequence databases are extremely redundant (think about gene homologs and genome sequencing of similar organisms or substrains of bacteria). CaBLAST restructures problems such that queries are exceuted against a precomputed, redundancy-minimized database. After the BLAST is peformed, resulting HSPs are compared against sequences within a the semi-redundant sequence cluster. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## Executing a BLAST Query through NCBI Entrez\n",
    "\n",
    "Given a protein of interest, one potential question is that we might want to figure out whether it has homolog of known structure. BLAST is well-suited for this problem!\n",
    "\n",
    "Let's look at the *E. coli* mechanosensitive channels to see which ones have known structures. We can [query](http://www.uniprot.org/uniprot/?query=mechanosensitive+AND+taxonomy%3A%22Escherichia+coli+%28strain+K12%29+%5B83333%5D%22&sort=score) Uniprot for proteins with \"mechanosensitive\" in their name and from the *E. coli* K12 genome.\n",
    "\n",
    "I've selected a subset here for our analysis.\n",
    "\n",
    "|Uniprot Accession| Entry Name | Description|\n",
    "|-----------------|------------| -----------|\n",
    "|[P0C0S1](http://www.uniprot.org/uniprot/P0C0S1)| MSCS_ECOLI | Small-conductance mechanosensitive channel |\n",
    "|[P0A742](http://www.uniprot.org/uniprot/P0A742)| MSCL_ECOLI | Large-conductance mechanosensitive channel |\n",
    "|[P39285](http://www.uniprot.org/uniprot/P39285)| MSCM_ECOLI | Miniconductance mechanosensitive channel |\n",
    "|[P77338](http://www.uniprot.org/uniprot/P77338)| MSCK_ECOLI | Mechanosensitive channel MscK (gated by K${}^{+}$)|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's first retrieve an example Genbank entry from NCBI corresponding to these accession number:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       MSCS_ECOLI               286 aa            linear   BCT 22-JUL-2015\n",
      "DEFINITION  RecName: Full=Small-conductance mechanosensitive channel.\n",
      "ACCESSION   P0C0S1\n",
      "VERSION     P0C0S1.1  GI:83288342\n",
      "DBSOURCE    UniProtKB: locus MSCS_ECOLI, accession P0C0S1;\n",
      "            class: standard.\n",
      "            extra accessions:P11666,Q2M9R8\n",
      "            created: Dec 6, 2005.\n",
      "            sequence updated: Dec 6, 2005.\n",
      "            annotation updated: Jul 22, 2015.\n",
      "            xrefs: X14436.1, CAA32606.1, U28377.1, AAA69091.1, U00096.3,\n",
      "            AAC75961.1, AP009048.1, BAE76988.1, QQEC4A, NP_417399.1,\n",
      "            WP_000389818.1, 2OAU_A, 2OAU_B, 2OAU_C, 2OAU_D, 2OAU_E, 2OAU_F,\n",
      "            2OAU_G, 2VV5_A, 2VV5_B, 2VV5_C, 2VV5_D, 2VV5_E, 2VV5_F, 2VV5_G,\n",
      "            3UDC_A, 3UDC_B, 3UDC_C, 3UDC_D, 3UDC_E, 3UDC_F, 3UDC_G, 4AGE_A,\n",
      "            4AGE_B, 4AGE_C, 4AGE_D, 4AGE_E, 4AGE_F, 4AGE_G, 4AGF_A, 4AGF_B,\n",
      "            4AGF_C, 4AGF_D, 4AGF_E, 4AGF_F, 4AGF_G, 4HWA_A, 4HWA_B, 4HWA_C,\n",
      "            4HWA_D, 4HWA_E, 4HWA_F, 4HWA_G\n",
      "            xrefs (non-sequence databases): PDBsum:2OAU, PDBsum:2VV5,\n",
      "            PDBsum:3UDC, PDBsum:4AGE, PDBsum:4AGF, PDBsum:4HWA,\n",
      "            ProteinModelPortal:P0C0S1, SMR:P0C0S1, DIP:DIP-36192N,\n",
      "            IntAct:P0C0S1, STRING:511145.b2924, TCDB:1.A.23.2.1, PaxDb:P0C0S1,\n",
      "            PRIDE:P0C0S1, EnsemblBacteria:AAC75961, EnsemblBacteria:AAC75961,\n",
      "            EnsemblBacteria:b2924, EnsemblBacteria:BAE76988,\n",
      "            EnsemblBacteria:BAE76988, EnsemblBacteria:BAE76988, GeneID:947416,\n",
      "            KEGG:eco:b2924, PATRIC:32121262, EchoBASE:EB1149, EcoGene:EG11160,\n",
      "            eggNOG:COG0668, HOGENOM:HOG000110050, InParanoid:P0C0S1, KO:K03442,\n",
      "            OMA:INIAYPQ, OrthoDB:EOG61GG69, PhylomeDB:P0C0S1,\n",
      "            BioCyc:EcoCyc:EG11160-MONOMER, BioCyc:ECOL316407:JW2891-MONOMER,\n",
      "            BioCyc:RETL1328306-WGS:GSTH-1410-MONOMER, EvolutionaryTrace:P0C0S1,\n",
      "            PRO:PR:P0C0S1, Proteomes:UP000000318, Proteomes:UP000000625,\n",
      "            GO:0016021, GO:0005886, GO:0008381, GO:0009992, GO:0034220,\n",
      "            InterPro:IPR010920, InterPro:IPR011066, InterPro:IPR006685,\n",
      "            InterPro:IPR006686, InterPro:IPR011014, InterPro:IPR008910,\n",
      "            Pfam:PF00924, Pfam:PF05552, SUPFAM:SSF50182, SUPFAM:SSF82689,\n",
      "            SUPFAM:SSF82861, PROSITE:PS01246\n",
      "KEYWORDS    3D-structure; Cell inner membrane; Cell membrane; Complete\n",
      "            proteome; Ion channel; Ion transport; Membrane; Reference proteome;\n",
      "            Transmembrane; Transmembrane helix; Transport.\n",
      "SOURCE      Escherichia coli K-12\n",
      "  ORGANISM  Escherichia coli K-12\n",
      "            Bacteria; Proteobacteria; Gammaproteobacteria; Enterobacteriales;\n",
      "            Enterobacteriaceae; Escherichia.\n",
      "REFERENCE   1  (residues 1 to 286)\n",
      "  AUTHORS   Alefounder,P.R. and Perham,R.N.\n",
      "  TITLE     Identification, molecular cloning and sequence analysis of a gene\n",
      "            cluster encoding the class II fructose 1,6-bisphosphate aldolase,\n",
      "            3-phosphoglycerate kinase and a putative second glyceraldehyde\n",
      "            3-phosphate dehydrogenase of Escherichia coli\n",
      "  JOURNAL   Mol. Microbiol. 3 (6), 723-732 (1989)\n",
      "   PUBMED   2546007\n",
      "  REMARK    NUCLEOTIDE SEQUENCE [GENOMIC DNA].;\n",
      "            STRAIN=K12 / CS520\n",
      "REFERENCE   2  (residues 1 to 286)\n",
      "  AUTHORS   Blattner,F.R., Plunkett,G. 3rd, Bloch,C.A., Perna,N.T., Burland,V.,\n",
      "            Riley,M., Collado-Vides,J., Glasner,J.D., Rode,C.K., Mayhew,G.F.,\n",
      "            Gregor,J., Davis,N.W., Kirkpatrick,H.A., Goeden,M.A., Rose,D.J.,\n",
      "            Mau,B. and Shao,Y.\n",
      "  TITLE     The complete genome sequence of Escherichia coli K-12\n",
      "  JOURNAL   Science 277 (5331), 1453-1462 (1997)\n",
      "   PUBMED   9278503\n",
      "  REMARK    NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].;\n",
      "            STRAIN=K12 / MG1655 / ATCC 47076\n",
      "REFERENCE   3  (residues 1 to 286)\n",
      "  AUTHORS   Hayashi,K., Morooka,N., Yamamoto,Y., Fujita,K., Isono,K., Choi,S.,\n",
      "            Ohtsubo,E., Baba,T., Wanner,B.L., Mori,H. and Horiuchi,T.\n",
      "  TITLE     Highly accurate genome sequences of Escherichia coli K-12 strains\n",
      "            MG1655 and W3110\n",
      "  JOURNAL   Mol. Syst. Biol. 2, 2006 (2006)\n",
      "   PUBMED   16738553\n",
      "  REMARK    NUCLEOTIDE SEQUENCE [LARGE SCALE GENOMIC DNA].;\n",
      "            STRAIN=K12 / W3110 / ATCC 27325 / DSM 5911\n",
      "REFERENCE   4  (residues 1 to 286)\n",
      "  AUTHORS   Martinac,B., Buechner,M., Delcour,A.H., Adler,J. and Kung,C.\n",
      "  TITLE     Pressure-sensitive ion channel in Escherichia coli\n",
      "  JOURNAL   Proc. Natl. Acad. Sci. U.S.A. 84 (8), 2297-2301 (1987)\n",
      "   PUBMED   2436228\n",
      "  REMARK    CHARACTERIZATION.;\n",
      "            STRAIN=K12 / AW405\n",
      "REFERENCE   5  (residues 1 to 286)\n",
      "  AUTHORS   Cui,C., Smith,D.O. and Adler,J.\n",
      "  TITLE     Characterization of mechanosensitive channels in Escherichia coli\n",
      "            cytoplasmic membrane by whole-cell patch clamp recording\n",
      "  JOURNAL   J. Membr. Biol. 144 (1), 31-42 (1995)\n",
      "   PUBMED   7595939\n",
      "  REMARK    SENSITIVITY TO OSMOTIC AND VOLTAGE DIFFERENCES ACROSS THE\n",
      "            CYTOPLASMIC MEMBRANE.;\n",
      "            STRAIN=K12\n",
      "REFERENCE   6  (residues 1 to 286)\n",
      "  AUTHORS   Levina,N., Totemeyer,S., Stokes,N.R., Louis,P., Jones,M.A. and\n",
      "            Booth,I.R.\n",
      "  TITLE     Protection of Escherichia coli cells against extreme turgor by\n",
      "            activation of MscS and MscL mechanosensitive channels:\n",
      "            identification of genes required for MscS activity\n",
      "  JOURNAL   EMBO J. 18 (7), 1730-1737 (1999)\n",
      "   PUBMED   10202137\n",
      "  REMARK    CHARACTERIZATION, AND PHYIOLOGICAL ROLE.;\n",
      "            STRAIN=K12 / MJF379\n",
      "REFERENCE   7  (residues 1 to 286)\n",
      "  AUTHORS   Sukharev,S.\n",
      "  TITLE     Purification of the small mechanosensitive channel of Escherichia\n",
      "            coli (MscS): the subunit structure, conduction, and gating\n",
      "            characteristics in liposomes\n",
      "  JOURNAL   Biophys. J. 83 (1), 290-298 (2002)\n",
      "   PUBMED   12080120\n",
      "  REMARK    LIPOSOME RECONSTITUTION.;\n",
      "            STRAIN=K12 / AW405\n",
      "REFERENCE   8  (residues 1 to 286)\n",
      "  AUTHORS   Okada,K., Moe,P.C. and Blount,P.\n",
      "  TITLE     Functional design of bacterial mechanosensitive channels.\n",
      "            Comparisons and contrasts illuminated by random mutagenesis\n",
      "  JOURNAL   J. Biol. Chem. 277 (31), 27682-27688 (2002)\n",
      "   PUBMED   12015316\n",
      "  REMARK    LIPOSOME RECONSTITUTION, AND MUTAGENESIS OF VAL-40.;\n",
      "            STRAIN=K12 / MJF379\n",
      "REFERENCE   9  (residues 1 to 286)\n",
      "  AUTHORS   Koprowski,P. and Kubalski,A.\n",
      "  TITLE     C termini of the Escherichia coli mechanosensitive ion channel\n",
      "            (MscS) move apart upon the channel opening\n",
      "  JOURNAL   J. Biol. Chem. 278 (13), 11237-11245 (2003)\n",
      "   PUBMED   12551944\n",
      "  REMARK    CROSS-LINKING OF THE C-TERMINI INHIBITS CHANNEL OPENING.;\n",
      "            STRAIN=K12 / MJF379\n",
      "REFERENCE   10 (residues 1 to 286)\n",
      "  AUTHORS   Miller,S., Edwards,M.D., Ozdemir,C. and Booth,I.R.\n",
      "  TITLE     The closed structure of the MscS mechanosensitive channel.\n",
      "            Cross-linking of single cysteine mutants\n",
      "  JOURNAL   J. Biol. Chem. 278 (34), 32246-32250 (2003)\n",
      "   PUBMED   12767977\n",
      "  REMARK    SUBUNIT IN THE CLOSED STATE, AND MUTAGENESIS OF SER-58 AND SER-267.\n",
      "REFERENCE   11 (residues 1 to 286)\n",
      "  AUTHORS   Schumann,U., Edwards,M.D., Li,C. and Booth,I.R.\n",
      "  TITLE     The conserved carboxy-terminus of the MscS mechanosensitive channel\n",
      "            is not essential but increases stability and activity\n",
      "  JOURNAL   FEBS Lett. 572 (1-3), 233-237 (2004)\n",
      "   PUBMED   15304354\n",
      "  REMARK    C-TERMINAL DELETIONS.\n",
      "REFERENCE   12 (residues 1 to 286)\n",
      "  AUTHORS   Stenberg,F., Chovanec,P., Maslen,S.L., Robinson,C.V., Ilag,L.L.,\n",
      "            von Heijne,G. and Daley,D.O.\n",
      "  TITLE     Protein complexes of the Escherichia coli cell envelope\n",
      "  JOURNAL   J. Biol. Chem. 280 (41), 34409-34419 (2005)\n",
      "   PUBMED   16079137\n",
      "  REMARK    SUBUNIT, AND SUBCELLULAR LOCATION.;\n",
      "            STRAIN=BL21-DE3\n",
      "REFERENCE   13 (residues 1 to 286)\n",
      "  AUTHORS   Daley,D.O., Rapp,M., Granseth,E., Melen,K., Drew,D. and von\n",
      "            Heijne,G.\n",
      "  TITLE     Global topology analysis of the Escherichia coli inner membrane\n",
      "            proteome\n",
      "  JOURNAL   Science 308 (5726), 1321-1323 (2005)\n",
      "   PUBMED   15919996\n",
      "  REMARK    TOPOLOGY [LARGE SCALE ANALYSIS].;\n",
      "            STRAIN=K12 / MG1655 / ATCC 47076\n",
      "REFERENCE   14 (residues 1 to 286)\n",
      "  AUTHORS   Edwards,M.D., Li,Y., Kim,S., Miller,S., Bartlett,W., Black,S.,\n",
      "            Dennison,S., Iscla,I., Blount,P., Bowie,J.U. and Booth,I.R.\n",
      "  TITLE     Pivotal role of the glycine-rich TM3 helix in gating the MscS\n",
      "            mechanosensitive channel\n",
      "  JOURNAL   Nat. Struct. Mol. Biol. 12 (2), 113-119 (2005)\n",
      "   PUBMED   15665866\n",
      "  REMARK    MUTAGENESIS OF CONSERVED GLY AND ALA RESIDUES IN TM3, AND\n",
      "            PRESENTATION OF MODELS OF THE CLOSED STATE CHANNEL AND CLOSED TO\n",
      "            OPEN STATE TRANSITIONS.;\n",
      "            STRAIN=K12 / MJF431\n",
      "REFERENCE   15 (residues 1 to 286)\n",
      "  AUTHORS   Bass,R.B., Strop,P., Barclay,M. and Rees,D.C.\n",
      "  TITLE     Crystal structure of Escherichia coli MscS, a voltage-modulated and\n",
      "            mechanosensitive channel\n",
      "  JOURNAL   Science 298 (5598), 1582-1587 (2002)\n",
      "   PUBMED   12446901\n",
      "  REMARK    X-RAY CRYSTALLOGRAPHY (3.9 ANGSTROMS) OF 27-280 PROBABLY IN THE\n",
      "            OPEN STATE, AND SUBUNIT.;\n",
      "            STRAIN=K12\n",
      "COMMENT     [FUNCTION] Mechanosensitive channel that participates in the\n",
      "            regulation of osmotic pressure changes within the cell, opening in\n",
      "            response to stretch forces in the membrane lipid bilayer, without\n",
      "            the need for other proteins. Forms an ion channel of 1.0\n",
      "            nanosiemens conductance with a slight preference for anions. The\n",
      "            channel is sensitive to voltage; as the membrane is depolarized,\n",
      "            less tension is required to open the channel and vice versa. The\n",
      "            channel is characterized by short bursts of activity that last for\n",
      "            a few seconds.\n",
      "            [FUNCTION] The channel pore is formed by TM3 and the loop between\n",
      "            TM2 and TM3. After a sharp turn at Gly-113, an alpha-helix\n",
      "            (residues 114-127) is oriented nearly parallel to the plane of the\n",
      "            putative lipid bilayer. On the intracellular side of the channel,\n",
      "            the permeation pathway of MscS does not connect directly to the\n",
      "            cytoplasm but instead opens to a large chamber that is connected to\n",
      "            the cytoplasm. This chamber resembles a molecular filter that could\n",
      "            serve to prescreen large molecules before they are allowed passage\n",
      "            to the transmembrane pore. The TM1 and TM2 helices appear to be\n",
      "            likely candidates for mediating the tension and voltage\n",
      "            sensitivities of MscS. Gating requires large rearrangements of at\n",
      "            least the C-terminus.\n",
      "            [SUBUNIT] Homoheptamer. {ECO:0000269|PubMed:12446901,\n",
      "            ECO:0000269|PubMed:12767977, ECO:0000269|PubMed:16079137}.\n",
      "            [SUBCELLULAR LOCATION] Cell inner membrane\n",
      "            {ECO:0000269|PubMed:16079137}; Multi-pass membrane protein\n",
      "            {ECO:0000269|PubMed:16079137}.\n",
      "            [SIMILARITY] Belongs to the MscS (TC 1.A.23) family. {ECO:0000305}.\n",
      "FEATURES             Location/Qualifiers\n",
      "     source          1..286\n",
      "                     /organism=\"Escherichia coli K-12\"\n",
      "                     /db_xref=\"taxon:83333\"\n",
      "     gene            1..286\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "     Protein         1..286\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /product=\"Small-conductance mechanosensitive channel\"\n",
      "                     /UniProtKB_evidence=\"Evidence at protein level\"\n",
      "     Region          1..286\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Mature chain\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"Small-conductance mechanosensitive channel.\n",
      "                     /FTId=PRO_0000110238.\"\n",
      "     Region          1..286\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"PRK10334\"\n",
      "                     /note=\"mechanosensitive channel MscS; Provisional\"\n",
      "                     /db_xref=\"CDD:182386\"\n",
      "     Region          1..28\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Topological domain\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"Periplasmic. {ECO:0000305|PubMed:15919996}.\"\n",
      "     Region          14..66\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"TM_helix\"\n",
      "                     /note=\"Conserved TM helix; pfam05552\"\n",
      "                     /db_xref=\"CDD:253249\"\n",
      "     Region          26..28\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Helical region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          29..57\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Transmembrane region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"Helical.\"\n",
      "     Region          30..59\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Helical region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Site            40\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /site_type=\"mutagenized\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"V->C,G,N: No detectable phenotype.\n",
      "                     {ECO:0000269|PubMed:12015316}.\"\n",
      "     Site            40\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /site_type=\"mutagenized\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"V->D,K: Normal growth stops, without cell death,\n",
      "                     due to increased membrane permeability to potassium ions\n",
      "                     and protons (permeability tested only for D\n",
      "                     substitutions). {ECO:0000269|PubMed:12015316}.\"\n",
      "     Region          58..67\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Topological domain\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"Cytoplasmic. {ECO:0000305|PubMed:15919996}.\"\n",
      "     Site            58\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /site_type=\"mutagenized\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"S->C: Readily forms disulfide bonds with\n",
      "                     cross-linkers, suggesting that individual S-58 are only 3\n",
      "                     Angstroms apart in the closed state, versus 33 Angstroms\n",
      "                     apart in the open state crystal structure.\n",
      "                     {ECO:0000269|PubMed:12767977}.\"\n",
      "     Region          63..87\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Helical region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          68..91\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Transmembrane region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"Helical.\"\n",
      "     Region          72..268\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"MS_channel\"\n",
      "                     /note=\"Mechanosensitive ion channel; pfam00924\"\n",
      "                     /db_xref=\"CDD:250229\"\n",
      "     Region          88..90\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Hydrogen bonded turn\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          92..95\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Topological domain\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"Periplasmic. {ECO:0000305|PubMed:15919996}.\"\n",
      "     Region          96..113\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Transmembrane region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"Helical.\"\n",
      "     Region          96..111\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Helical region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          113..125\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Helical region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          114..286\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Topological domain\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"Cytoplasmic. {ECO:0000305|PubMed:15919996}.\"\n",
      "     Region          135..140\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Beta-strand region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          142..148\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Beta-strand region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          150..156\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Beta-strand region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          162..166\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Beta-strand region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          167..171\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Helical region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          175..192\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Beta-strand region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          198..211\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Helical region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          221..228\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Beta-strand region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          230..242\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Beta-strand region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          243..245\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Hydrogen bonded turn\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Region          246..264\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Helical region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:2VV5}.\"\n",
      "     Site            266..286\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /site_type=\"mutagenized\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"ISFPYPQMDVNFKRVKEDKAA->HHHHHHLE: Normal levels of\n",
      "                     channels are expressed; they recover more slowly than\n",
      "                     wild-type cells after desensitization.\n",
      "                     {ECO:0000269|PubMed:15665866}.\"\n",
      "     Site            266..286\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /site_type=\"mutagenized\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"ISFPYPQMDVNFKRVKEDKAA->LE: Fewer channels present\n",
      "                     in the membrane, they require slightly more pressure to\n",
      "                     open and do not recover after desensitization.\n",
      "                     {ECO:0000269|PubMed:15665866}.\"\n",
      "     Site            267\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /site_type=\"mutagenized\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"S->C: Provides biochemical evidence for heptameric\n",
      "                     structure upon cross-linking.\n",
      "                     {ECO:0000269|PubMed:12767977}.\"\n",
      "     Region          274..279\n",
      "                     /gene=\"mscS\"\n",
      "                     /locus_tag=\"b2924\"\n",
      "                     /gene_synonym=\"JW2891\"\n",
      "                     /gene_synonym=\"yggB\"\n",
      "                     /region_name=\"Beta-strand region\"\n",
      "                     /experiment=\"experimental evidence, no additional details\n",
      "                     recorded\"\n",
      "                     /note=\"{ECO:0000244|PDB:3UDC}.\"\n",
      "ORIGIN      \n",
      "        1 medlnvvdsi ngagswlvan qalllsyavn ivaalaiiiv gliiarmisn avnrlmisrk\n",
      "       61 idatvadfls alvrygiiaf tliaalgrvg vqtasviavl gaaglavgla lqgslsnlaa\n",
      "      121 gvllvmfrpf rageyvdlgg vagtvlsvqi fsttmrtadg kiivipngki iagniinfsr\n",
      "      181 epvrrnefii gvaydsdidq vkqiltniiq sedrilkdre mtvrlnelga ssinfvvrvw\n",
      "      241 snsgdlqnvy wdvlerikre fdaagisfpy pqmdvnfkrv kedkaa\n",
      "//\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Always tell NCBI who you are\n",
    "Bio.Entrez.email = 'saladi@caltech.edu'\n",
    "Bio.Entrez.tool = 'BLAST-for-PDB-Entry'\n",
    "\n",
    "id_list = ('P0C0S1', 'P0A742', 'P39285', 'P77338')\n",
    "handle_list = {}\n",
    "\n",
    "# example record retrieval\n",
    "handle = Bio.Entrez.efetch(db='protein', id=id_list[0], rettype='gb', retmode='text')\n",
    "print(handle.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, let's retrieve the records for each of the identifiers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'P0A742': SeqRecord(seq=Seq('MSIIKEFREFAMRGNVVDLAVGVIIGAAFGKIVSSLVADIIMPPLGLLIGGIDF...NRS', IUPACProtein()), id='P0A742.1', name='MSCL_ECOLI', description='RecName: Full=Large-conductance mechanosensitive channel.', dbxrefs=[]), 'P77338': SeqRecord(seq=Seq('MTMFQYYKRSRHFVFSAFIAFVFVLLCQNTAFARASSNGDLPTKADLQAQLDSL...AVG', IUPACProtein()), id='P77338.1', name='MSCK_ECOLI', description='RecName: Full=Mechanosensitive channel MscK; AltName: Full=Potassium efflux system KefA; Flags: Precursor.', dbxrefs=[]), 'P39285': SeqRecord(seq=Seq('MRLIITFLMAWCLSWGAYAATAPDSKQITQELEQAKAAKPAQPEVVEALQSALN...GSL', IUPACProtein()), id='P39285.3', name='MSCM_ECOLI', description='RecName: Full=Miniconductance mechanosensitive channel MscM; Flags: Precursor.', dbxrefs=[]), 'P0C0S1': SeqRecord(seq=Seq('MEDLNVVDSINGAGSWLVANQALLLSYAVNIVAALAIIIVGLIIARMISNAVNR...KAA', IUPACProtein()), id='P0C0S1.1', name='MSCS_ECOLI', description='RecName: Full=Small-conductance mechanosensitive channel.', dbxrefs=[])}\n"
     ]
    }
   ],
   "source": [
    "record_list = {}\n",
    "\n",
    "# Fetch all the handles, read each into a seq_record\n",
    "for thisid in id_list:\n",
    "    handle = Bio.Entrez.efetch(db='protein', id=thisid, rettype='gb', retmode='text')\n",
    "    seq_record = Bio.SeqIO.read(handle, 'genbank')\n",
    "    handle.close()\n",
    "    record_list[thisid] = seq_record\n",
    "    \n",
    "\n",
    "print(record_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given the sequence, Biopython has a method to query NCBI's Web Blast service that returns an XML file. There are a huge number of options! Read more about each in the documentation [here](http://biopython.org/DIST/docs/api/Bio.Blast.NCBIWWW-module.html).\n",
    "\n",
    ">`qblast(program, \n",
    "       database, \n",
    "       sequence, \n",
    "       auto_format=None, \n",
    "       composition_based_statistics=None, \n",
    "       db_genetic_code=None, \n",
    "       endpoints=None, \n",
    "       entrez_query='(none)', \n",
    "       expect=10.0, \n",
    "       filter=None, \n",
    "       gapcosts=None, \n",
    "       genetic_code=None, \n",
    "       hitlist_size=50, \n",
    "       i_thresh=None, \n",
    "       layout=None, \n",
    "       lcase_mask=None, \n",
    "       matrix_name=None, \n",
    "       nucl_penalty=None, \n",
    "       nucl_reward=None, \n",
    "       other_advanced=None, \n",
    "       perc_ident=None, \n",
    "       phi_pattern=None, \n",
    "       query_file=None, \n",
    "       query_believe_defline=None, \n",
    "       query_from=None, \n",
    "       query_to=None, \n",
    "       searchsp_eff=None, \n",
    "       service=None, \n",
    "       threshold=None, \n",
    "       ungapped_alignment=None, \n",
    "       word_size=None, \n",
    "       alignments=500,\n",
    "       alignment_view=None, \n",
    "       descriptions=500, \n",
    "       entrez_links_new_window=None,\n",
    "       expect_low=None, \n",
    "       expect_high=None, \n",
    "       format_entrez_query=None, \n",
    "       format_object=None, \n",
    "       format_type='XML',\n",
    "       ncbi_gi=None, \n",
    "       results_file=None, \n",
    "       show_overview=None,\n",
    "       megablast=None)`\n",
    "       \n",
    "If you find that `Bio.Blast.NCBIWWW.qblast()` does not give the same results as the NCBI BLAST website, it is likely becuase the default search parameters/options are likely different between the web version and qblast. Check parameters like gap penalties and expectation threshold.\n",
    "\n",
    "We are only required to specify the program, database, and sequence to run a simple query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "result_handle = Bio.Blast.NCBIWWW.qblast('blastp', 'pdb', record_list['P77338'].seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['P0A742', 'P77338', 'P39285', 'P0C0S1'])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record_list.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I like to save the results of my BLAST query to file before I parse them in case I need to go back and start the parsing over."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "with open('blastp-pdb-P77338.xml', 'w') as save_file:\n",
    "    save_file.write(result_handle.read())\n",
    "\n",
    "result_handle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check that the file was created ok:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "!head blastp-pdb-P77338.xml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's do this for the entire set of identifiers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# perform all blast queries and write each to file\n",
    "for thisid in record_list:\n",
    "    # perform blast query\n",
    "    result_handle = Bio.Blast.NCBIWWW.qblast('blastp', 'pdb', record_list[thisid].seq)\n",
    "    \n",
    "    # write query to file\n",
    "    filename = \"blast-pdb-{recordid}.xml\".format(recordid=thisid)\n",
    "    with open(filename, 'w+') as save_file:\n",
    "        save_file.write(result_handle.read())\n",
    "\n",
    "    result_handle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Biopython provides two ways to parse the output of BLAST queries. One is specalized for BLAST-XML output [`Bio.Blast.NCBIXML`](http://biopython.org/DIST/docs/api/Bio.Blast.NCBIXML-module.html) and the other, [`Bio.SearchIO`](http://biopython.org/DIST/docs/api/Bio.SearchIO-module.html) is more general for all sorts of sequence search methods including BLAST but also for other methods, e.g. [HMMER](http://hmmer.janelia.org/) and [Exonerate](http://www.ebi.ac.uk/about/vertebrate-genomics/software/exonerate). While Biopython issues a warning about using `Bio.SearchIO` since it's \"an experimental submodule\", it's been stable for 2 years and any further changes should be minor.\n",
    "\n",
    "We will use it here since the same code could be adapted for future work with other search methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!ls *.xml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Program: blastp (2.2.32+)\n",
      "  Query: unnamed (1120)\n",
      "         protein product\n",
      " Target: pdb\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  gi|873090862|pdb|5AJI|A  Chain A, Mscs D67r1 High Resol...\n",
      "            1      1  gi|400977297|pdb|4AGE|A  Chain A, Mtssl Spin Labeled D6...\n",
      "            2      1  gi|126031460|pdb|2OAU|A  Chain A, Mechanosensitive Chan...\n",
      "            3      1  gi|400977304|pdb|4AGF|A  Chain A, Mtssl Spin Labeled L1...\n",
      "            4      1  gi|195927344|pdb|2VV5|A  Chain A, The Open Structure Of...\n",
      "            5      1  gi|449802644|pdb|4HW9|A  Chain A, Crystal Structure Of ...\n",
      "            6      1  gi|410562602|pdb|3UDC|A  Chain A, Crystal Structure Of ...\n",
      "            7      1  gi|410562584|pdb|3T9N|A  Chain A, Crystal Structure Of ...\n",
      "            8      1  gi|399124994|pdb|3ZUG|A  Chain A, E268d Mutant Of Fad S...\n",
      "            9      1  gi|747155358|pdb|4U0V|A  Chain A, Crystal Structure Of ...\n",
      "           10      1  gi|283806768|pdb|2WV0|A  Chain A, Crystal Structure Of ...\n"
     ]
    }
   ],
   "source": [
    "blast_qresult = Bio.SearchIO.read('blastp-pdb-P77338.xml', 'blast-xml')\n",
    "print(blast_qresult)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At first glance, it looks like there are many hits of proteins with known structure with homology to MscK. Let's parse these hits to determine whether the hits are bonafide -- do they span the whole MscK protein sequence or only one/few domains?\n",
    "\n",
    "To do so, let's examine the structure of `Bio.SearchIO` objects. `SearchIO` objects hold multiple `hit` objects each of which holds a `hsp` (\"High Scoring Pair\") objects. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      Query: unnamed protein product\n",
      "        Hit: gi|873090862|pdb|5AJI|A Chain A, Mscs D67r1 High Resolution\n",
      "Query range: [884:1102] (0)\n",
      "  Hit range: [68:284] (0)\n",
      "Quick stats: evalue 1.9e-17; bitscore 85.11\n",
      "  Fragments: 1 (224 columns)\n",
      "     Query - ITTILNYIIIAVGAMTVFGSLGVSWDKLQWLAAALSVGLGFGLQEIFGNFVSGLIILFE~~~EKGDE\n",
      "             ++ ++ Y IIA   +   G +GV    +  +  A  + +G  LQ    N  +G++++  ~~~ K D+\n",
      "       Hit - LSALVRYGIIAFTLIAALGRVGVQTASVIAVLGAAGLAVGLALQGSLSNLAAGVLLVMF~~~VKEDK\n",
      "      Query: unnamed protein product\n",
      "        Hit: gi|400977297|pdb|4AGE|A Chain A, Mtssl Spin Labeled D67c Mutant ...\n",
      "Query range: [884:1102] (0)\n",
      "  Hit range: [68:284] (0)\n",
      "Quick stats: evalue 2e-17; bitscore 85.11\n",
      "  Fragments: 1 (224 columns)\n",
      "     Query - ITTILNYIIIAVGAMTVFGSLGVSWDKLQWLAAALSVGLGFGLQEIFGNFVSGLIILFE~~~EKGDE\n",
      "             ++ ++ Y IIA   +   G +GV    +  +  A  + +G  LQ    N  +G++++  ~~~ K D+\n",
      "       Hit - LSALVRYGIIAFTLIAALGRVGVQTASVIAVLGAAGLAVGLALQGSLSNLAAGVLLVMF~~~VKEDK\n",
      "      Query: unnamed protein product\n",
      "        Hit: gi|126031460|pdb|2OAU|A Chain A, Mechanosensitive Channel Of Sma...\n",
      "Query range: [884:1102] (0)\n",
      "  Hit range: [88:304] (0)\n",
      "Quick stats: evalue 2.9e-17; bitscore 84.73\n",
      "  Fragments: 1 (224 columns)\n",
      "     Query - ITTILNYIIIAVGAMTVFGSLGVSWDKLQWLAAALSVGLGFGLQEIFGNFVSGLIILFE~~~EKGDE\n",
      "             ++ ++ Y IIA   +   G +GV    +  +  A  + +G  LQ    N  +G++++  ~~~ K D+\n",
      "       Hit - LSALVRYGIIAFTLIAALGRVGVQTASVIAVLGAAGLAVGLALQGSLSNLAAGVLLVMF~~~VKEDK\n",
      "      Query: unnamed protein product\n",
      "        Hit: gi|400977304|pdb|4AGF|A Chain A, Mtssl Spin Labeled L124c Mutant...\n",
      "Query range: [884:1102] (0)\n",
      "  Hit range: [68:284] (0)\n",
      "Quick stats: evalue 3.2e-17; bitscore 84.34\n",
      "  Fragments: 1 (224 columns)\n",
      "     Query - ITTILNYIIIAVGAMTVFGSLGVSWDKLQWLAAALSVGLGFGLQEIFGNFVSGLIILFE~~~EKGDE\n",
      "             ++ ++ Y IIA   +   G +GV    +  +  A  + +G  LQ    N  +G++ +  ~~~ K D+\n",
      "       Hit - LSALVRYGIIAFTLIAALGRVGVQTASVIAVLGAAGLAVGLALQGSLSNLAAGVLCVMF~~~VKEDK\n",
      "      Query: unnamed protein product\n",
      "        Hit: gi|195927344|pdb|2VV5|A Chain A, The Open Structure Of Mscs\n",
      "Query range: [884:1102] (0)\n",
      "  Hit range: [68:284] (0)\n",
      "Quick stats: evalue 6e-17; bitscore 83.57\n",
      "  Fragments: 1 (224 columns)\n",
      "     Query - ITTILNYIIIAVGAMTVFGSLGVSWDKLQWLAAALSVGLGFGLQEIFGNFVSGLIILFE~~~EKGDE\n",
      "             ++ ++ Y IIA   +   G +GV    +  +  A  + +G  LQ    N  +G++++  ~~~ K D+\n",
      "       Hit - LSALVRYGIIAFTLIAALGRVGVQTASVIAVLGAAGLVVGLALQGSLSNLAAGVLLVMF~~~VKEDK\n",
      "      Query: unnamed protein product\n",
      "        Hit: gi|449802644|pdb|4HW9|A Chain A, Crystal Structure Of Helicobact...\n",
      "Query range: [901:1114] (0)\n",
      "  Hit range: [98:307] (0)\n",
      "Quick stats: evalue 3.5e-10; bitscore 63.16\n",
      "  Fragments: 1 (219 columns)\n",
      "     Query - FGSLGVSWDKLQWLAAALSVGLGFGLQEIFGNFVSGLIILFERPVRIGDTVTIGSFSGT~~~YKGDD\n",
      "               +LGV    +  +   + + +   L++   +   G+I++   P + GD + I    G ~~~YK DD\n",
      "       Hit - LSTLGVQTTSIITVLGTVGIAVALALKDYLSSIAGGIILIILHPFKKGDIIEISGLEGK~~~YKDDD\n",
      "      Query: unnamed protein product\n",
      "        Hit: gi|410562602|pdb|3UDC|A Chain A, Crystal Structure Of A Membrane...\n",
      "Query range: [888:1018] (0)\n",
      "  Hit range: [71:197] (0)\n",
      "Quick stats: evalue 8.1e-10; bitscore 61.62\n",
      "  Fragments: 1 (130 columns)\n",
      "     Query - LNYIIIAVGAMTVFGSLGVSWDKLQWLAAALSVGLGFGLQEIFGNFVSGLIILFERPVR~~~DLEKV\n",
      "             + YII  +   ++     +    L  +A   S+ +GFG Q +  + +SG  I+FE    ~~~D++K+\n",
      "       Hit - VRYIIYFLAGASILKLFNIDMTSLLAVAGIGSLAIGFGAQNLVKDMISGFFIIFEDQFS~~~DVDKI\n",
      "      Query: unnamed protein product\n",
      "        Hit: gi|410562584|pdb|3T9N|A Chain A, Crystal Structure Of A Membrane...\n",
      "Query range: [888:1018] (0)\n",
      "  Hit range: [71:197] (0)\n",
      "Quick stats: evalue 8.4e-10; bitscore 61.62\n",
      "  Fragments: 1 (130 columns)\n",
      "     Query - LNYIIIAVGAMTVFGSLGVSWDKLQWLAAALSVGLGFGLQEIFGNFVSGLIILFERPVR~~~DLEKV\n",
      "             + YII  +   ++     +    L  +A   S+ +GFG Q +  + +SG  I+FE    ~~~D++K+\n",
      "       Hit - VRYIIYFLAGASILKLFNIDMTSLLAVAGIGSLAIGFGAQNLVKDMISGFFIIFEDQFS~~~DVDKI\n",
      "      Query: unnamed protein product\n",
      "        Hit: gi|399124994|pdb|3ZUG|A Chain A, E268d Mutant Of Fad Synthetase ...\n",
      "Query range: [30:107] (0)\n",
      "  Hit range: [246:319] (0)\n",
      "Quick stats: evalue 3.1; bitscore 31.96\n",
      "  Fragments: 1 (79 columns)\n",
      "     Query - AFARASSNGDLPTKADLQAQLDS--LNKQKDLSAQDKLVQQDLTDTLATLDKIDRIKEE~~~RQATA\n",
      "             A+A A S G  PT  D Q  +DS  L++  DL   D  V+ +  D +  ++K D ++  ~~~R   A\n",
      "       Hit - AYAAAISVGTNPTFGDEQRSVDSFVLDRDADLYGHD--VKVEFVDHVRAMEKFDSVE--~~~RTLLA\n",
      "      Query: unnamed protein product\n",
      "        Hit: gi|747155358|pdb|4U0V|A Chain A, Crystal Structure Of Yvoa From ...\n",
      "Query range: [989:1052] (0)\n",
      "  Hit range: [105:172] (0)\n",
      "Quick stats: evalue 4.3; bitscore 31.19\n",
      "  Fragments: 1 (67 columns)\n",
      "     Query - RLINWSLTDTTTRLVIRLGVAYGSDLEKVRKVLLK----AATEHPRVMHEPMPEVFFTAFGASTLDH\n",
      "             RLI++ L D+T  L   LG  + S + K+ +V L      A E   +  E   E+  + F +S  DH\n",
      "       Hit - RLIDYQLIDSTEELAAILGCGHPSSIHKITRVRLANDIPMAIESSHIPFELAGELNESHFQSSIYDH\n",
      "      Query: unnamed protein product\n",
      "        Hit: gi|283806768|pdb|2WV0|A Chain A, Crystal Structure Of The Gntr-H...\n",
      "Query range: [989:1052] (0)\n",
      "  Hit range: [102:169] (0)\n",
      "Quick stats: evalue 4.5; bitscore 31.19\n",
      "  Fragments: 1 (67 columns)\n",
      "     Query - RLINWSLTDTTTRLVIRLGVAYGSDLEKVRKVLLK----AATEHPRVMHEPMPEVFFTAFGASTLDH\n",
      "             RLI++ L D+T  L   LG  + S + K+ +V L      A E   +  E   E+  + F +S  DH\n",
      "       Hit - RLIDYQLIDSTEELAAILGCGHPSSIHKITRVRLANDIPMAIESSHIPFELAGELNESHFQSSIYDH\n"
     ]
    }
   ],
   "source": [
    "for blast_hit in blast_qresult:\n",
    "    for hsp in blast_hit:\n",
    "        print(hsp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's focus in on the query range, hit range, and statistics of the HSP to evaluate whether we have *real* BLAST hits for each of the query sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID\t\tlength\t\thit_start\t\thit_end\t\thit_span\t\tquery_start\t\tquery_end\t\tquery_span\n",
      "P0A742\t\t136\t\t20\t\t114\t\t94\t\t2\t\t106\t\t104\n",
      "P77338\t\t1120\t\t68\t\t284\t\t216\t\t884\t\t1102\t\t218\n",
      "P39285\t\t1107\t\t28\t\t278\t\t250\t\t828\t\t1084\t\t256\n",
      "P0C0S1\t\t286\t\t20\t\t306\t\t286\t\t0\t\t286\t\t286\n"
     ]
    }
   ],
   "source": [
    "print(\"ID\", \"length\", \"hit_start\", \"hit_end\", \"hit_span\",\n",
    "        \"query_start\", \"query_end\", \"query_span\", sep=\"\\t\\t\")\n",
    "\n",
    "for thisid in record_list:\n",
    "    filename = \"blastp-pdb-{recordid}.xml\".format(recordid=thisid)\n",
    "    thisresult = Bio.SearchIO.read(filename, 'blast-xml')\n",
    "    best_hsp = thisresult[0][0]\n",
    "    print(thisid, len(record_list[thisid]), best_hsp.hit_start, best_hsp.hit_end, best_hsp.hit_span,\n",
    "          best_hsp.query_start, best_hsp.query_end, best_hsp.query_span, sep=\"\\t\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
